core-examples/rust/message-monitoring/src/main.rs

```diff
@@ -0,0 +1,245 @@
+        Err(_) => KeyPair::new(
+            "2ada2e65ab8eeab09490e3521415f45b6e42df9c760a639bcf53957550b25a16".to_owned(),
+            "172af540e43a524763dd53b26a066d472a97c4de37d5498170564510608250c3".to_owned(),
+        ),
+    };
+    giver_keys.secret.truncate(64);
+
+    let mut messages = Vec::with_capacity(args.message_count as usize);
+
+    for i in 0..args.message_count {
+        let value: u64 = (i as u64 + 1) * 1_000_000_000;
+        let msg = encode_message(
+            context.clone(),
+            ParamsOfEncodeMessage {
+                abi: Abi::Contract(from_value::<AbiContract>(giver_abi.clone()).unwrap()),
+                address: Some(address.clone()),
+                call_set: CallSet::some_with_function_and_input(
+                    "sendTransaction",
+                    json!({
+                        "dest": address,
+                        "value": dbg!(value),
+                        "bounce": false,
+                    }),
+                ),
+                signer: Signer::Keys {
+                    keys: giver_keys.clone(),
+                },
+                ..Default::default()
+            },
+        )
+        .await
+        .unwrap();
+        dbg!(msg.message_id);
+        messages.push(MessageSendingParams {
+            boc: msg.message,
+            wait_until,
+            ..Default::default()
+        });
+    }
+
+    let snd = send_messages(
+        context.clone(),
+        ParamsOfSendMessages {
+            messages,
+            monitor_queue: Some(QUEUE.to_owned()),
+            ..Default::default()
+        },
+    )
+    .await
+    .unwrap();
+
+    dbg!(snd.messages.len());
+
+    if args.await_all {
+        let fetch_res = fetch_next_monitor_results(
+            context.clone(),
+            ParamsOfFetchNextMonitorResults {
+                queue: QUEUE.to_owned(),
+                wait: Some(MonitorFetchWait::All),
+                ..Default::default()
+            },
+        )
+        .await
+        .unwrap();
+
+        dbg!(fetch_res.results);
+    }
+
+    if args.once_await {
+        let mut fetched_results = Vec::with_capacity(args.message_count as usize);
+
+        loop {
+            let fetch_res = fetch_next_monitor_results(
+                context.clone(),
+                ParamsOfFetchNextMonitorResults {
+                    queue: QUEUE.to_owned(),
+                    wait: Some(MonitorFetchWait::AtLeastOne),
+                    ..Default::default()
+                },
+            )
+            .await
+            .unwrap();
+
+            dbg!(fetch_res.results.len());
+
+            // If monitor return empty results this indicates that all messages resolved and fetched
+            if fetch_res.results.len() == 0 {
+                break;
+            }
+
+            for item in fetch_res.results {
+                fetched_results.push(item);
+            }
+
+            sleep(Duration::from_secs(LOOP_SLEEP_IN_SECONDS)).await;
```

Comment:
"AtLeastOnes" it is "Subscribe" in other words.
The semantics of "subscribe" suggests that you want to know about events as soon as possible, so `sleep` here looks unnatural.